### ISSUE TITLE
JWT.Encode with certificate private key is not working

### ISSUE THREAD

**Issue by zzggs** [dvsekhvalnov/jose-jwt#181]
We are using following sample code:
`var payload = new Dictionary<string, object>()
{
    { "sub", "mr.x@contoso.com" },
    { "exp", 1300819380 }
};

var headers = new Dictionary<string, object>()
{
     { "typ", "JWT" },
     { "cty", "JWT" },
     { "keyid", "111-222-333"}
};

var privateKey=new X509Certificate2("my-key.p12", "password", X509KeyStorageFlags.Exportable | X509KeyStorageFlags.MachineKeySet).PrivateKey as RSACryptoServiceProvider;

string token=Jose.JWT.Encode(payload, privateKey, JwsAlgorithm.RS256, extraHeaders: headers);`

It seems no signature in this token although token format is correct as "BASE64URL(UTF8(JWS Protected Header)).BASE64URL(JWS Payload).BASE64URL(JWS Signature)", when we use JwtSecurityTokenHandler to read this token:
`var handler = new JwtSecurityTokenHandler();
var tokenData = handler.ReadJwtToken(token);`

![image](https://user-images.githubusercontent.com/104475425/165433143-f39aa060-224a-499a-8627-b41520c41d80.png)

What's wrong with the code?

**Comment by dvsekhvalnov** [comment]
Hi @zzggs , 

can you post `token` value after executing:

``` cs
string token=Jose.JWT.Encode(payload, privateKey, JwsAlgorithm.RS256, extraHeaders: headers);
```
?

I really can't comment on `JwtSecurityTokenHandler`, have no idea what's that.

**Comment by zzggs** [comment]
> 
Hi @dvsekhvalnov ,

Thanks a lot for your reply.

I found the issue by myself, the issue is related to free managed certificate on Azure app serive, seems it can not be used to sign and encrypt. I will try a new SSL certificate instead.

I put token on [https://jwt.io/](url) , it told me invalid signature. I take a lot time to research source code and Java JWS and JWE object, finally I'm in wrong direction.

**Comment by dvsekhvalnov** [comment]
So do you have more questions or help? Or good to close issue?

**Comment by zzggs** [comment]
> So do you have more questions or help? Or good to close issue?

No, we created a self-signed SSL certificate, same issue throwed.

I put token on [https://jwt.io/](https://github.com/dvsekhvalnov/jose-jwt/issues/url) , it also told me invalid signature.

Refer to sample code:
`var payload = new Dictionary<string, object>()
{
    { "sub", "mr.x@contoso.com" },
    { "exp", 1300819380 }
};

var privateKey=new X509Certificate2("my-key.p12", "password").GetRSAPrivateKey();

string token=Jose.JWT.Encode(payload, privateKey, JwsAlgorithm.RS256);`

**Comment by dvsekhvalnov** [comment]
1. Did you put your key to jwt.io? It will always tell you wrong signature unless you add your verification key.
2. See my last comment here https://github.com/dvsekhvalnov/jose-jwt/issues/179 is it helpful ?

**Comment by zzggs** [comment]
> 1. Did you put your key to jwt.io? It will always tell you wrong signature unless you add your verification key.
> 2. See my last comment here [AES/GCM/NoPadding decryption failed: Tag mismatch! #179](https://github.com/dvsekhvalnov/jose-jwt/issues/179) is it helpful ?

Hi dvsekhvalnov,

Thanks for your help, I have read the post you mentioned, still cannot find the reason.

I'm using JWT.Encode twice to generate signed token with our private key and encrypted token with public key from another cert.

I also try using JWK as parameter to generated signed token:
`Jwk encryptionKey = new Jwk(privatekey, isPrivate: true);
var jwksign = Jose.JWT.Encode(payload, encryptionKey, JwsAlgorithm.RS256, extraHeaders: header);`

payload is our custom information, header contains a kid in. Seems nothing is different, still get Tag mismatch error.

**Comment by zzggs** [comment]
> 
Hi dvsekhvalnov,

I have modified my previous post, code is there, please help me check code. Seems nothing is incorrect.

Do you mean receiver side is using a different cert to verify our token? Seems this is also possible.

**Comment by dvsekhvalnov** [comment]
What's inside `header` and `header1`?   

Do you have receiver code as well? 

Yes, i meant receiver side using different cert and authn tag didn't match, so it can't decrypt.

**Comment by zzggs** [comment]
> What's inside `header` and `header1`?
> 
> Do you have receiver code as well?
> 
> Yes, i meant receiver side using different cert and authn tag didn't match, so it can't decrypt.

I have checked with receiver, they are using correct cert.

Header and header1 only contain one key with different kid, also will include alg after encoding.

Still not found what's wrong with my code.

Response : AES/GCM/NoPadding decryption failed: Tag mismatch!

**Comment by dvsekhvalnov** [comment]
Show full snippet so i can run it and try against Nimbus. I'm assuming receiver using NimbusDS java library?

**Comment by zzggs** [comment]
> Show full snippet so i can run it and try against Nimbus. I'm assuming receiver using NimbusDS java library?

Hi dvsekhvalnov,

Yes, you are great. Receiver is using Java Nimbus-JOSE-JWT. I only got Java sample code, no .Net sample.

I can share Java sample code:
**Sign method**
private String decryptMessage(String respMsgPayload, KeyStoreFactory keyStore)
  throws KeyStoreException, NoSuchAlgorithmException, CertificateException, IOException, 
         java.text.ParseException, UnrecoverableKeyException, JOSEException {
    JWEObject jweObject = JWEObject.parse(respMsgPayload);
            
    PrivateKey privateKey = (PrivateKey) keyStore.getPrivateKey("merchant_private_key_alias");
            
    JWEDecrypter decrypter = new RSADecrypter(privateKey);
  jweObject.decrypt(decrypter);
            
    String signedMessage = jweObject.getPayload().toString();
    return signedMessage;
}

**Encrypt Method**
private String verifySignature(String signedMessage, KeyStore ks, String keyAlias)
  throws KeyStoreException, JOSEException, ParseException {
    JWSObject jwsObject = JWSObject.parse(signedMessage);
            
    Certificate certificate = ks.getCertificate(keyAlias);
    JWSVerifier verifier = new RSASSAVerifier((RSAPublicKey) certificate.getPublicKey());
            
    if (!jwsObject.verify(verifier)) {
      throw new ValidationException("Invalid Signature");
    }
    return jwsObject.getPayload().toString();
}

**Comment by dvsekhvalnov** [comment]
Your .net code is fine. While your java snippet is not receiver, it doing same thing as .NET sign and encrypt.

Can you show java part that decoding your .net created token?

**Comment by zzggs** [comment]
> Your .net code is fine. While your java snippet is not receiver, it doing same thing as .NET sign and encrypt.
> 
> Can you show java part that decoding your .net created token?
I'm afraid the answer is no. Receiver just show sample code, I had contacted them before, they told us we need to sign message with our private key first. There must be different bewteen Java and .Net code. It seems following code makes different change:
`jwsObject.sign(signer);`

I think this is not a technology issue, I have no idea, so take a chance here.

**Comment by dvsekhvalnov** [comment]
Well, it's hard for me to help further without additional info. It sounds like mismatch with keys to me, probably dig into given direction. Did you register your public key at receiver?

**Comment by zzggs** [comment]
> AES/GCM/NoPadding decryption failed: Tag mismatch!

Yes, we confirmed with receiver before. Receiver told us the message could be decrypted just now, I will try again to check if it is working now. Thanks.
